File: morphir-dotnet/src/Morphir.CodeModel/IR/QName.fs
module Morphir.IR.QName

open Morphir.IR.Name
open Morphir.IR.Path
open Morphir.SDK

type QName = QName of modulePath: Path * localName: Name

let qName (modulePath: Path) (localName: Name) = QName(modulePath, localName)

/// Turn a qualified name into a tuple.
let toTuple (QName(path, name) as qName) = (path, name)

/// Turn a tuple into a qualified name.
let fromTuple =
    function
    | (path, name) -> QName(path, name)

/// Create a qualified name.
let fromName modulePath localName = QName(modulePath, localName)

/// Get the module path part of a qualified name.
let getModulePath (QName(modulePath, _) as qName) = modulePath

/// Get the local name part of a qualified name.
let getLocalName (QName(_, localName) as qName) = localName

/// Turn a QName into a string using ':' as a separator between module and local names.
let toString (QName(moduleName, localName) as qName) =
    String.join ":" [ Path.toString Name.toTitleCase "." moduleName; Name.toCamelCase localName ]

File: morphir/src/org/finos/morphir/ir/gen1/qname.scala
package org.finos.morphir.ir.gen1

/// A qualified name (`QName`) is a combination of a module path and a local name.
sealed case class QName(moduleName: ModuleName, localName: Name) {
  def modulePath: Path = moduleName.path

  /// Turn a qualified name into a tuple of a module path and a local name.
  @inline def toTuple: (Path, Name) = (modulePath, localName)

  override def toString: String =
    modulePath.toString(Name.toTitleCase, ".") + ":" + localName.toCamelCase

}

object QName {
  val empty: QName = QName(ModuleName.empty, Name.empty)

  def apply(moduleName: String, localName: String): QName =
    QName(ModuleName.fromString(moduleName), Name.fromString(localName))

  /// Turn a qualified name into a tuple of a module path and a local name.
  def toTuple(qName: QName): (Path, Name) = qName.toTuple

  /// Turn a tuple of a module path and a local name into a qualified name (`QName`).
  def fromTuple(tuple: (Path, Name)): QName = QName(ModuleName(tuple._1), tuple._2)

  /// Creates a qualified name from a module path and a local name.
  def fromName(modulePath: Path, localName: Name): QName = QName(ModuleName(modulePath), localName)

  /// Creates a qualified name from strings representing a module path and a local name.
  def fromName(modulePath: String, localName: String): QName =
    QName(ModuleName(Path.fromString(modulePath)), Name.fromString(localName))

  /// Get the local name part of a qualified name.
  def getLocalName(qname: QName): Name = qname.localName

  /// Get the module path part of a qualified name.
  def getModulePath(qname: QName): Path = qname.modulePath

  /// Turn a `QName` into a string using `:` as the separator between the module path and the local name.
  def toString(qName: QName): String = qName.toString

  /// Parse a string into a qualified name using `:` as the separator between the module path and the local name.
  def fromString(str: String): Option[QName] =
    str.split(":") match {
      case Array(packageNameString, localNameString) =>
        Some(QName(ModuleName(Path.fromString(packageNameString)), Name.fromString(localNameString)))
      case _ => None
    }
}
